burgerSync.js

```diff
@@ -0,0 +1,252 @@
+                break; 
+            case this.MESSAGE_TYPE.INITIAL_HANDSHAKE_RESPONSE: 
+                this._printMessage('Initial handshake received!');
+                if (message.cumulativeDifficulty > this.burgerNode.info.cumulativeDifficulty) {
+                    this._printMessage('Peer chain is better, requesting sync...');
+                    this._write(peer, this.MESSAGE_TYPE.REQUEST_SYNC_CHAIN, 'Let us sync!');
+                }
+                break;
+            case this.MESSAGE_TYPE.REQUEST_SYNC_CHAIN:
+                this._printMessage('Sync request received! Sending chain...')
+                this._write(peer, this.MESSAGE_TYPE.RESPONSE_SYNC_CHAIN, this.burgerNode.chain);
+                break;
+            case this.MESSAGE_TYPE.RESPONSE_SYNC_CHAIN:
+                /**
+                 * When chain is synced, all pending transactions are lost.
+                 * Wait for Issue #41 and adjust needed logic to concat
+                 * pending transactions instead of replacing.
+                 */
+                this._printMessage('Chain received! Syncing...')
+                this.burgerNode.replaceChain(message);
+                this._printMessage('    Synced!');
+                break;
+            case this.MESSAGE_TYPE.BROADCAST_NEW_BLOCK:
+                this._printMessage('A new block has been mined!');
+                if (message.cumulativeDifficulty > this.burgerNode.info.cumulativeDifficulty) {
+                    this.burgerNode.replaceChain(message);
+                    this.broadcastNewBlock(message);
+                }
+                break;
+            case this.MESSAGE_TYPE.BROADCAST_NEW_TRANSACTION:
+                this._printMessage('Processing new broadcasted transaction...');
+                const isValid = this.burgerNode.addPendingTransaction(message);
+                if (isValid) {
+                    this.burgerNode.addPendingTransaction(message);
+                    this._broadcast(this.MESSAGE_TYPE.BROADCAST_NEW_TRANSACTION, message);
+                    this._printMessage('    Added to pending transactions!');
+                } else {
+                    this._printMessage('    Not added to chain!')
+                }
+                break;
+            default:
+                this._write(peer, this.MESSAGE_TYPE.INVALID_REQUEST, 'REJECTED: Got invalid request!')
+        }
+    }
+
+    /**
+     * Broadcasts a new transaction to all connected peers.
+     * 
+     * @param {burgerTransaction} transaction 
+     */
+    broadcastNewTransaction(transaction) {
+        this._broadcast(this.MESSAGE_TYPE.BROADCAST_NEW_TRANSACTION, transaction);
+    }
+
+    /**
+     * Broadcasts a new block to all connected peers.
+     * This event happens when a new block is mined.
+     * 
+     * @param {burgerBlockchain} chain 
+     */
+    broadcastNewBlock(chain) {
+        this._broadcast(this.MESSAGE_TYPE.BROADCAST_NEW_BLOCK, chain);
```

Comment:
@24thsaint I don't know how I feel about broadcasting a new block by sending over the entire chain...

Reply:
I agree. This should be replaced when we implement an append block function somewhere.